bzzeth/bzzeth.go

```diff
@@ -99,6 +119,215 @@ func (b *BzzEth) handleMsgFromSwarmNode(p *Peer) func(context.Context, interface
 	}
 }
 
+// handles new header hashes - strategy; only request headers that are in Kad Nearest Neighbourhood
+func (b *BzzEth) handleNewBlockHeaders(ctx context.Context, p *Peer, msg *NewBlockHeaders) {
+	p.logger.Debug("bzzeth.handleNewBlockHeaders")
+
+	// collect the addresses of blocks that are not in our localstore
+	var addresses []chunk.Address
+	for _, h := range *msg {
+		addresses = append(addresses, h.Hash.Bytes())
+		log.Trace("Received hashes ", "Header", hex.EncodeToString(h.Hash.Bytes()))
+	}
+	yes, err := b.netStore.Store.HasMulti(ctx, addresses...)
+
+	// collect the hashes of block headers we want
+	var hashes [][]byte
+	for i, y := range yes {
+		// ignore hashes already present in localstore
+		if y {
+			continue
+		}
+
+		// collect hash based on proximity
+		vhash := addresses[i]
+		if wantHeaderFunc(vhash, b.kad) {
+			hashes = append(hashes, vhash)
+		} else {
+			p.logger.Trace("ignoring header. Not in proximity ", "Address", hex.EncodeToString(addresses[i]))
+		}
+	}
+
+	// request them from the offering peer and deliver in a channel
+	deliveries := make(chan []byte)
+	req, err := p.getBlockHeaders(ctx, hashes, deliveries)
+	if err != nil {
+		p.logger.Error("Error sending GetBlockHeader message", "Reason", err)
+		return
+	}
+	defer req.cancel()
+
+	// this loop blocks until all delivered or context done
+	// only needed to log results
+	deliveredCnt := 0
+	for {
+		select {
+		case hash, ok := <-deliveries:
+			if !ok {
+				p.logger.Debug("bzzeth.handleNewBlockHeaders", "hash", hex.EncodeToString(hash), "delivered", deliveredCnt)
+				return
+			}
+			deliveredCnt++
+			p.logger.Trace("bzzeth.handleNewBlockHeaders", "hash", hex.EncodeToString(hash), "delivered", deliveredCnt)
+			if deliveredCnt == len(req.hashes) {
+				p.logger.Debug("Delivered all headers", "count", deliveredCnt)
+				finishDeliveryFunc(req.hashes)
+				return
+			}
+		case <-ctx.Done():
+			p.logger.Debug("bzzeth.handleNewBlockHeaders", "delivered", deliveredCnt, "err", err)
+			return
+		}
+	}
+}
+
+// wantHeaderFunc is used to determine if we need a particular header offered as latest
+// by an eth fullnode
+// tests reassign this to control
+var wantHeaderFunc = wantHeader
+
+// wantHeader returns true iff the hash argument falls in the NN of kademlia
+func wantHeader(hash []byte, kad *network.Kademlia) bool {
+	return chunk.Proximity(kad.BaseAddr(), hash) >= kad.NeighbourhoodDepth()
+}
+
+// finishStorageFunc is used to determine if all the requested headers are stored
+// this is used in testing if the headers are indeed stored in the localstore
+var finishStorageFunc = finishStorage
+
+func finishStorage(chunks []chunk.Chunk) {
+	for _, c := range chunks {
```

Comment:
I understand this is needed for testing (although this pluggable function can probably be avoided using `SubscribePull` from the test in some way), but shouldn't we print the hashes anyway? I would inline the content of all of these pluggable functions to the place they are used from, and just check `if finishStorageFunc != nil { finishStorageFunc(chunks) }`. this is a bit quirky at the moment IMO

Reply:
Well.. printing hashes is not needed in testing.

I dont see why this is Quirky...